File: 253_meeting_rooms_2.cpp
// O(NlogN)
class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        // O(NlogN)
        sort(intervals.begin(), intervals.end(), 
             [](const vector<int>& interval1, const vector<int>& interval2) {
                return interval1[0] < interval2[0];
        });
        
        // O(NlogR)
        int n = intervals.size();
        priority_queue<int, vector<int>, greater<int>> pq;
        
        for(int i=0; i<n; i++)
        {
            if(!pq.empty() && pq.top() <= intervals[i][0])
            {
                pq.pop();
            }
            pq.push(intervals[i][1]);
        }
        
        return pq.size();
    }
};

File: 253_meeting_rooms_2.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        intervals.sort(key= lambda i: i[0])
        if not intervals:
            return 0
        free_rooms = []
        heapq.heappush(free_rooms, intervals[0][1])
        for i in intervals[1:]:
            if free_rooms[0] <= i[0]:
                heapq.heappop(free_rooms)

            heapq.heappush(free_rooms, i[1])

        return len(free_rooms)